serving/src/main/java/feast/auth/credentials/JwtAuthCredentials.java

```diff
@@ -0,0 +1,110 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package feast.auth.credentials;
+
+import static io.grpc.Metadata.ASCII_STRING_MARSHALLER;
+
+import io.grpc.CallCredentials2;
+import io.grpc.Metadata;
+import io.grpc.Status;
+import java.util.Map;
+import java.util.concurrent.Executor;
+import javax.security.sasl.AuthenticationException;
+import okhttp3.FormBody;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import org.springframework.boot.configurationprocessor.json.JSONObject;
+
+/*
+ * Oauth Credentials Implementation for serving.
+ * Used by CoreSpecService to connect to core.
+ */
+public class JwtAuthCredentials extends CallCredentials2 {
+
+  static final String APPLICATION_JSON = "application/json";
+  static final String CONTENT_TYPE = "content-type";
+  static final String BEARER_TYPE = "Bearer";
+  static final String GRANT_TYPE = "grant_type";
+  static final String CLIENT_ID = "client_id";
+  static final String CLIENT_SECRET = "client_secret";
+  static final String AUDIENCE = "audience";
+  static final String OAUTH_URL = "oauth_url";
+  static final Metadata.Key<String> AUTHORIZATION_METADATA_KEY =
+      Metadata.Key.of("Authorization", ASCII_STRING_MARSHALLER);
+
+  private OkHttpClient httpClient;
+  private Request request;
+
+  public JwtAuthCredentials(CoreAuthenticationProperties authProperties) {
+    this.httpClient = new OkHttpClient();
+    Map<String, String> options = authProperties.getOptions();
+    if (!(options.containsKey(GRANT_TYPE)
+        && options.containsKey(CLIENT_ID)
+        && options.containsKey(AUDIENCE)
+        && options.containsKey(CLIENT_SECRET)
+        && options.containsKey(OAUTH_URL))) {
+      throw new AssertionError(
+          "please configure the properties:"
+              + " grant_type, client_id, client_secret, audience, oauth_url");
+    }
+    RequestBody requestBody =
+        new FormBody.Builder()
+            .add(GRANT_TYPE, options.get(GRANT_TYPE))
+            .add(CLIENT_ID, options.get(CLIENT_ID))
+            .add(CLIENT_SECRET, options.get(CLIENT_SECRET))
+            .add(AUDIENCE, options.get(AUDIENCE))
+            .build();
+    this.request =
+        new Request.Builder()
+            .url(options.get(OAUTH_URL))
+            .addHeader(CONTENT_TYPE, APPLICATION_JSON)
+            .post(requestBody)
+            .build();
+  }
+
+  @Override
+  public void thisUsesUnstableApi() {
+    // TODO Auto-generated method stub
+
+  }
+
+  @Override
+  public void applyRequestMetadata(
+      RequestInfo requestInfo, Executor appExecutor, MetadataApplier applier) {
+    appExecutor.execute(
+        new Runnable() {
+          @Override
+          public void run() {
+            try {
+              Response response = httpClient.newCall(request).execute();
```

Comment:
Does this mean that for each incoming request we will have an outgoing request for a token?

Reply:
access tokens need to be cached until near expiry, if not done already.  @jmelinav let me know if you need an example.

Reply:
Right now yes, I will push an update with caching.